--- /usr/local/share/ghostscript/6.51/lib/Fontmap.GS.orig	Fri Aug 24 16:49:36 2001
+++ /usr/local/share/ghostscript/6.51/lib/Fontmap.GS	Fri Aug 24 17:03:57 2001
@@ -140,10 +140,10 @@
 /Bookman-Light			/URWBookmanL-Ligh	;
 /Bookman-LightItalic		/URWBookmanL-LighItal	;
 
-/Courier			/NimbusMonL-Regu	;
-/Courier-Oblique		/NimbusMonL-ReguObli	;
-/Courier-Bold			/NimbusMonL-Bold	;
-/Courier-BoldOblique		/NimbusMonL-BoldObli	;
+% /Courier			/NimbusMonL-Regu	;
+% /Courier-Oblique		/NimbusMonL-ReguObli	;
+% /Courier-Bold			/NimbusMonL-Bold	;
+% /Courier-BoldOblique		/NimbusMonL-BoldObli	;
 
 /AvantGarde-Book		/URWGothicL-Book	;
 /AvantGarde-BookOblique		/URWGothicL-BookObli	;
@@ -170,10 +170,10 @@
 /NewCenturySchlbk-Bold		/CenturySchL-Bold	;
 /NewCenturySchlbk-BoldItalic	/CenturySchL-BoldItal	;
 
-/Times-Roman			/NimbusRomNo9L-Regu	;
-/Times-Italic			/NimbusRomNo9L-ReguItal	;
-/Times-Bold			/NimbusRomNo9L-Medi	;
-/Times-BoldItalic		/NimbusRomNo9L-MediItal	;
+% /Times-Roman			/NimbusRomNo9L-Regu	;
+% /Times-Italic			/NimbusRomNo9L-ReguItal	;
+% /Times-Bold			/NimbusRomNo9L-Medi	;
+% /Times-BoldItalic		/NimbusRomNo9L-MediItal	;
 
 /Symbol				/StandardSymL	;
 
@@ -404,3 +404,50 @@
 % the SymbolEncoding.
 
 /Hershey-Symbol				(hrsyr.gsf)	;	% 5066567
+
+% Cyrrillic koi8 fonts/aliases
+
+/ArialCyrMT			(/usr/local/share/ps-cyr/arial8.pfb)	;
+/ArialCyrMT-Bold		(/usr/local/share/ps-cyr/arialbd8.pfb)	;
+/ArialCyrMT-Italic		(/usr/local/share/ps-cyr/ariali8.pfb)	;
+/ArialCyrMT-BoldItalic		(/usr/local/share/ps-cyr/arialbi8.pfb)	;
+
+/Arial			/ArialCyrMT			;
+/Arial-Bold		/ArialCyrMT-Bold		;
+/Arial-Italic		/ArialCyrMT-Italic		;
+/Arial-BoldItalic	/ArialCyrMT-BoldItalic		;
+
+/Helvetica		/ArialCyrMT			;
+/Helvetica-Bold		/ArialCyrMT-Bold		;
+/Helvetica-Oblique	/ArialCyrMT-Italic		;
+/Helvetica-BoldOblique	/ArialCyrMT-BoldItalic		;
+
+/TimesNRCyrMT			(/usr/local/share/ps-cyr/times8.pfb)	;
+/TimesNRCyrMT-Bold		(/usr/local/share/ps-cyr/timesbd8.pfb)	;
+/TimesNRCyrMT-Inclined		(/usr/local/share/ps-cyr/timesi8.pfb)	;
+/TimesNRCyrMT-BoldInclined	(/usr/local/share/ps-cyr/timesbi8.pfb)	;
+
+/Times-Roman		/TimesNRCyrMT			;
+/Times-Bold		/TimesNRCyrMT-Bold		;
+/Times-Italic		/TimesNRCyrMT-Inclined		;
+/Times-BoldItalic	/TimesNRCyrMT-BoldInclined	;
+
+/CourierCyrPS			(/usr/local/share/ps-cyr/cour8.pfb)	;
+/CourierCyrPS-Bold		(/usr/local/share/ps-cyr/courbd8.pfb)	;
+/CourierCyrPS-Inclined		(/usr/local/share/ps-cyr/couri8.pfb)	;
+/CourierCyrPS-BoldInclined	(/usr/local/share/ps-cyr/courbi8.pfb)	;
+
+/ERKurierKOI8RRegular		(/usr/local/share/ps-cyr/cokoi8n.pfb)	;
+/ERKurierKOI8RBold		(/usr/local/share/ps-cyr/cokoi8b.pfb)	;
+/ERKurierKOI8RItalic		(/usr/local/share/ps-cyr/cokoi8i.pfb)	;
+/ERKurierKOI8RBoldItalic 	(/usr/local/share/ps-cyr/cokoi8bi.pfb)	;
+
+% /Courier		/CourierCyrPS			;
+% /Courier-Bold		/CourierCyrPS-Bold		;
+% /Courier-Oblique	/CourierCyrPS-Inclined		;
+% /Courier-BoldOblique	/CourierCyrPS-BoldInclined	;
+
+/Courier		/ERKurierKOI8RRegular		;
+/Courier-Bold		/ERKurierKOI8RBold		;
+/Courier-Oblique	/ERKurierKOI8RItalic		;
+/Courier-BoldOblique	/ERKurierKOI8RBoldItalic	;
