$RuOBSD: ru-plain-koi8-r.patch,v 1.1 2006/06/28 16:35:01 form Exp $
--- minimalist.orig	Wed Jun 28 23:53:08 2006
+++ minimalist	Wed Jun 28 23:53:15 2006
@@ -419,6 +419,26 @@
    if ($subject !~ /^(.*:\s+)?\[$list\] /i) {
      $subject = "[$list] ".$subject; }
   }
+
+ $header =~ /(^|\n)Content-Type:[ \t]+(.*\n([ \t]+.*\n)*)/i;
+ $ctyped = $2;
+ if ($language eq "ru") {
+   if ($ctyped) {
+     $ctyped =~ /charset="?(.*?)"?[;\s]/i;
+     $msgcharset = lc($1);
+   }
+   if (!$ctyped || $ctyped !~ /text\/plain/ || $msgcharset ne "koi8-r") {
+     $msg = <<_EOF_ ;
+To: $mailto
+Subject: $subject
+
+ОШИБКА:
+Вы отправляете письмо в недопустимом формате или в неправильной кодировке.
+Допускается только формат text/plain, кодировка koi8-r.
+_EOF_
+     goto SendMessage;
+   }
+ }
   
  open LIST, "list" and do {
    while ($ent = <LIST>) {
