Index: files.ata
===================================================================
RCS file: /data/cvs/src/sys/dev/ata/files.ata,v
retrieving revision 1.1
diff -u -r1.1 files.ata
--- files.ata	18 Jul 1999 21:25:17 -0000	1.1
+++ files.ata	19 May 2002 09:57:59 -0000
@@ -12,4 +12,8 @@
 file	dev/ata/wd.c				wd needs-flag
 file	dev/ata/ata_wdc.c			wd & wdc_base
 
+device	ad: disk
+attach	ad at ata
+file	dev/ata/ad.c				ad needs-flag
+
 file	dev/ata/ata.c				ata | atapi
