Index: files.pci
===================================================================
RCS file: /cvs/src/sys/dev/pci/files.pci,v
retrieving revision 1.149
diff -u -r1.149 files.pci
--- files.pci	2003/08/19 03:36:21	1.149
+++ files.pci	2003/11/12 16:02:50
@@ -427,6 +427,11 @@
 attach	gtp at pci
 file	dev/pci/gtp.c			gtp
 
+# Pcimax Ultra FM-Transmitter
+device	pcmax: radiobus
+attach	pcmax at pci
+file	dev/pci/pcmax_pci.c		pcmax
+
 # National Semiconductor DP83820/DP83821 based GigE
 device	nge: ether, ifnet, mii, ifmedia, mii_phy
 attach	nge at pci
