Index: conf.h
===================================================================
RCS file: /cvs/src/sys/sys/conf.h,v
retrieving revision 1.48
diff -u -r1.48 conf.h
--- conf.h	2001/09/28 02:53:13	1.48
+++ conf.h	2001/10/03 06:40:10
@@ -416,6 +416,13 @@
 	(dev_type_stop((*))) enodev, 0, (dev_type_select((*))) enodev, \
 	(dev_type_mmap((*))) enodev }
 
+/* open, close, ioctl */
+#define       cdev_radio_init(c,n) { \
+	dev_init(c,n,open), dev_init(c,n,close), (dev_type_read((*))) enodev, \
+	(dev_type_write((*))) enodev, dev_init(c,n,ioctl), \
+	(dev_type_stop((*))) enodev, 0, (dev_type_select((*))) enodev, \
+	(dev_type_mmap((*))) enodev }
+
 /* symbolic sleep message strings */
 extern char devopn[], devio[], devwait[], devin[], devout[];
 extern char devioc[], devcls[];
@@ -490,7 +497,7 @@
 cdev_decl(audio);
 cdev_decl(midi);
 cdev_decl(sequencer);
-
+cdev_decl(radio);
 cdev_decl(cn);
 
 bdev_decl(sw);
