$RuOBSD: joe.patch,v 1.4 2008/10/01 21:53:12 form Exp $

--- rc/ftyperc.orig	Fri Oct 31 07:42:29 2008
+++ rc/ftyperc	Mon Feb  2 10:45:33 2009
@@ -1,8 +1,6 @@
  File Type table.  This is :included by the joerc file.
 
- No '.' in file name?  Assume it's a text file and we want wordwrap on.
 *
--wordwrap
 
  File name with '.'?  It's probably not a text file.
 *.*
@@ -44,149 +42,105 @@
 
  Troff
 *.1
--wordwrap
 -syntax troff
 
 *.1.in
--wordwrap
 -syntax troff
 
  TeX
 *.tex
--wordwrap
 -syntax tex
 -tex_comment
 -no_double_quoted
 
 *.sty
--wordwrap
 -syntax tex
 -tex_comment
 -no_double_quoted
 
  Text file.
 *.txt
--wordwrap
 
  News/mail files.
 *.article*
--wordwrap
 -syntax mail
 
 *.followup
--wordwrap
 -syntax mail
 
 *.letter
--wordwrap
 -syntax mail
 
  NN newsreader
 *tmp/nn.*
--wordwrap
 -syntax mail
 
  mail
 *tmp/Re*
--wordwrap
 -syntax mail
 
  elm
 *tmp/snd.*
--wordwrap
 -syntax mail
 
  dmail
 *tmp/dmt*
--wordwrap
 -syntax mail
 
  pine
 *tmp/pico.*
--wordwrap
 -syntax mail
 
  Assembly language
 *.asm
--wordwrap
 -syntax asm
 
 *.s
--wordwrap
 -syntax asm
 
 *.S
--wordwrap
 -syntax asm
 
  Mason
 *.mas
--autoindent
 -syntax mason
--smarthome
--smartbacks
 
  SML
 *.sml
--autoindent
 -syntax sml
--smarthome
--smartbacks
--istep 2
 
  OCaml lanaguage file
 *.ml
--autoindent
 -syntax ocaml
--smarthome
--smartbacks
--istep 2
 
  OCaml language interface
 *.mli
--autoindent
 -syntax ocaml
--smarthome
--smartbacks
--istep 2
 
  Ruby
 *.rb
--autoindent
 -syntax ruby
 
  Perl
 *
 +#!\+\[ 	]\+\[a-z/]/perl\>
--autoindent
 -syntax perl
--smarthome
--smartbacks
 -pound_comment
 -single_quoted
 
 *.pl
--autoindent
 -syntax perl
--smarthome
--smartbacks
 -pound_comment
 -single_quoted
 
 *.pm
--autoindent
 -syntax perl
--smarthome
--smartbacks
 -pound_comment
 -single_quoted
 
  SQL file
 *.sql
--autoindent
 -syntax sql
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -197,19 +151,13 @@
 
  AWK language file
 *.awk
--autoindent
 -syntax awk
--smarthome
--smartbacks
 -purify
 -pound_comment
 
  YACC
 *.y
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -219,10 +167,7 @@
 
  LEX
 *.l
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -231,10 +176,7 @@
 -cpara >#!;*/%
 
 *.lex
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -245,20 +187,12 @@
  ADA
 *.adb
 -syntax ada
--autoindent
--istep 2
--smarthome
--smartbacks
 -purify
 -vhdl_comment
 -text_delimiters declare|Declare|DECLARE|exception|Exception|EXCEPTION|if|If|IF|loop|Loop|LOOP|case|Case|CASE|package|Package|PACKAGE|procedure|Procedure|PROCEDURE|record|Record|RECORD|function|Function|FUNCTION=end|End|END
 
 *.ads
 -syntax ada
--autoindent
--istep 2
--smarthome
--smartbacks
 -purify
 -vhdl_comment
 -text_delimiters declare|Declare|DECLARE|exception|Exception|EXCEPTION|if|If|IF|loop|Loop|LOOP|case|Case|CASE|package|Package|PACKAGE|procedure|Procedure|PROCEDURE|record|Record|RECORD|function|Function|FUNCTION=end|End|END
@@ -273,16 +207,13 @@
  REXX
 *.rex
 -syntax rexx
--autoindent
 
 *.REX
 -syntax rexx
--autoindent
 
 *
 +#!\+\[ 	]\+\[a-z/]/rxx\>
 -syntax rexx
--autoindent
 
  sed
 *.sed
@@ -297,10 +228,7 @@
 
  C language file
 *.c
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -309,10 +237,7 @@
 -cpara >#!;*/%
 
 *.cpp
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -321,10 +246,7 @@
 -cpara >#!;*/%
 
 *.cc
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -333,10 +255,7 @@
 -cpara >#!;*/%
 
 *.c++
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -346,10 +265,7 @@
 
  C language header file
 *.h
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -358,10 +274,7 @@
 -cpara >#!;*/%
 
 *.h.in
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -370,10 +283,7 @@
 -cpara >#!;*/%
 
 *.hpp
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -382,10 +292,7 @@
 -cpara >#!;*/%
 
 *.h++
--autoindent
 -syntax c
--smarthome
--smartbacks
 -purify
 -single_quoted
 -c_comment
@@ -395,11 +302,7 @@
 
  Verilog file
 *.v
--autoindent
 -syntax verilog
--istep 2
--smarthome
--smartbacks
 -purify
 -c_comment
 -cpp_comment
@@ -408,11 +311,7 @@
 
  Verilog header file
 *.vh
--autoindent
 -syntax verilog
--istep 2
--smarthome
--smartbacks
 -purify
 -c_comment
 -cpp_comment
@@ -421,11 +320,7 @@
 
  VHDL file
 *.vhd
--autoindent
 -syntax vhdl
--istep 2
--smarthome
--smartbacks
 -purify
 -vhdl_comment
 -text_delimiters entity|ENTITY|if|IF|component|COMPONENT|loop|LOOP|configuration|CONFIGURATION|units|UNITS|record|RECORD|case|CASE|function|FUNCTION|package|PACKAGE|architecture|ARCHITECTURE|block|BLOCK|process|PROCESS|generate|GENERATE=end|END
@@ -433,124 +328,91 @@
 
  XML
 *.xml
--wordwrap
--autoindent
 -syntax xml
 
  XML
 *.xsd
--wordwrap
--autoindent
 -syntax xml
 
  HTML
 *.htm
--wordwrap
--autoindent
 -syntax html
 
 *.html
--wordwrap
--autoindent
 -syntax html
 
  JAVA
 *.java
--autoindent
 -syntax java
--smarthome
--smartbacks
 -purify
 -cpara >#!;*/%
 
  Javascript
 *.js
--autoindent
 -syntax java
--smarthome
--smartbacks
 -purify
 -cpara >#!;*/%
 
  Hypertext preprocessor file
 *.php
--autoindent
 -syntax php
--smarthome
--smartbacks
 -purify
 
  Python
 *
 +#!\+\[ 	]\+\[a-z/]\+\[ 	]python\>
--autoindent
 -syntax python
--smarthome
--smartbacks
 -purify
 
 *.py
--autoindent
 -syntax python
--smarthome
--smartbacks
 -purify
 
  CSH
 *
 +#!\+\[ 	]\+\[a-z/]/csh\>
--autoindent
 -syntax csh
 -pound_comment
 
 *.csh
--autoindent
 -syntax csh
 -pound_comment
 
 *.login
--autoindent
 -syntax csh
 -pound_comment
 
 *.logout
--autoindent
 -syntax csh
 -pound_comment
 
  TCSH
 *
 +#!\+\[ 	]\+\[a-z/]/tcsh\>
--autoindent
 -syntax csh
 -pound_comment
 
 *.tcsh
--autoindent
 -syntax csh
 -pound_comment
 
 *.tcshrc
--autoindent
 -syntax csh
 -pound_comment
 
  Shell
 *
 +#!\+\[ 	]\+\[a-z/]/sh\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.sh
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *profile
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
@@ -558,161 +420,125 @@
  BASH Shell
 *
 +#!\+\[ 	]\+\[a-z/]/bash\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash_login
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash_logout
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
  LISP
 *.lisp
--autoindent
 -syntax lisp
 -semi_comment
 
 *.lsp
--autoindent
 -syntax lisp
 -semi_comment
 
 *.el
--autoindent
 -syntax lisp
 -semi_comment
 
  KSH
 *
 +#!\+\[ 	]\+\[a-z/]/ksh\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.ksh
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
  Makefile
 *akefile
--autoindent
 -syntax conf
 -pound_comment
 
 *akefile.in
--autoindent
 -syntax conf
 -pound_comment
 
 *AKEFILE
--autoindent
 -syntax conf
 -pound_comment
 
  Pascal
 *.p
--autoindent
 -syntax pascal
--smarthome
--smartbacks
 -purify
 -text_delimiters begin|BEGIN|record|RECORD|case|CASE=end|END:repeat|REPEAT=until|UNTIL
 
 *.pas
--autoindent
 -syntax pascal
--smarthome
--smartbacks
 -purify
 -text_delimiters begin|BEGIN|record|RECORD|case|CASE=end|END:repeat|REPEAT=until|UNTIL
 
  Fortran
 *.f
--autoindent
 -syntax fortran
 
 *.for
--autoindent
 -syntax fortran
 
 *.FOR
--autoindent
 -syntax fortran
 
  Cadence SKILL
 *.il
 -semi_comment
 -syntax skill
--autoindent
--istep 3
 -indentc 32
--smartbacks
--smarthome
 -indentfirst
 
  LUA
 *.lua
--autoindent
 -syntax lua
 
  TCL
 *.tcl
--autoindent
 -syntax tcl
 
  Joe Syntax File
 *.jsf
--autoindent
 -syntax jsf
--smarthome
 
 
 *.jsf.in
--autoindent
 -syntax jsf
--smarthome
 
  Autoconfig file
 *.ac
--autoindent
 -syntax m4
 
  M4 file
 *.m4
--autoindent
 -syntax m4
 
  Matlab file
 *.m
--wordwrap
 -syntax matlab
 
  Automake file
 *.am
--autoindent
 -syntax conf
 
  Mail file
 *tmp/mutt-*
--wordwrap
 -syntax mail
 
  System RC file
--- rc/joerc.in.orig	Sun Nov  2 09:29:43 2008
+++ rc/joerc.in	Mon Feb  2 10:47:58 2009
@@ -61,14 +61,14 @@
 
  -force		Force final newline when files are saved
 
- -nolocks	If you don't want file locks to be used
+-nolocks	If you don't want file locks to be used
 
  -nomodcheck	Disable periodic checking of disk file newer than buffer
 		(this checking happens on save even with this option).
 
  -nocurdir	Do not prompt with current directory
 
- -nobackups	If you don't want backup files to be created
+-nobackups	If you don't want backup files to be created
 
  -break_hardlinks
 		Delete file before writing, to break hard links
@@ -123,7 +123,7 @@
 
  -skiptop nnn	Don't use top nnn lines of the screen
 
--notite         Don't send terminal initialization and termination strings:
+ -notite	Don't send terminal initialization and termination strings:
                 prevents restoration of screen on exit.
 
  -usetabs       Use tabs to help optimize screen update.
@@ -162,7 +162,7 @@
 
  -autoswap	Swap markb with markk when necessary
 
--joe_state     Use ~/.joe_state file
+ -joe_state	Use ~/.joe_state file
 
  -mouse		Enable xterm mouse support.  When enabled, left-click will
 		position the cursor and left-click-drag will select blocks
@@ -192,7 +192,7 @@
 		For example: -text_color bg_blue+white
 		Sets the background color to blue and foreground color to white.
 
--restore	Restore previous cursor position when files are opened
+ -restore	Restore previous cursor position when files are opened
 
 -search_prompting
 		Search prompts with previous search request.
@@ -367,7 +367,7 @@
 
  Default local options
 -highlight
--istep 2
+ -istep 2
 
  Perforce macro: use this macro (put in first column) to have joe "p4 edit"
  a file you're about to change.
