$RuOBSD$

--- b.c.orig	Tue Nov 15 19:50:20 2005
+++ b.c	Tue Nov 15 19:50:28 2005
@@ -260,8 +260,10 @@
 			close (b->filehandle);
 		if (b->changed)
 			abrerr(b->name);
+#ifndef NO_DOTLOCK
 		if (b->locked && !b->ignored_lock && plain_file(b))
 			unlock_it(b->name);
+#endif
 		if (b == errbuf)
 			errbuf = NULL;
 		if (b->undo)
--- cmd.c.orig	Tue Nov 15 19:50:43 2005
+++ cmd.c	Tue Nov 15 19:54:16 2005
@@ -245,6 +245,7 @@
 	if (c=='s' || c=='S') {
 		unsigned char bf1[256];
 		unsigned char bf[300];
+#ifndef NO_DOTLOCK
 		unlock_it(b->name);
 		if (lock_it(b->name,bf1)) {
 			int x;
@@ -262,6 +263,9 @@
 				return -1;
 			}
 		} else {
+#else
+		{
+#endif
 			b->locked=1;
 			if (notify)
 				*notify = 1;
@@ -303,6 +307,7 @@
 		unsigned char bf1[256];
 		unsigned char bf[300];
 		int x;
+#ifndef NO_DOTLOCK
 		/* It's a plain file- try to lock it */
 		if (lock_it(b->name,bf1)) {
 			for(x=0;bf1[x] && bf1[x]!=':';++x);
@@ -318,6 +323,9 @@
 			} else
 				return 0;
 		} else {
+#else
+			{
+#endif
 			/* Remember to unlock it */
 			b->locked = 1;
 		}
--- joerc.in.orig	Tue Nov 15 20:21:49 2005
+++ joerc.in	Tue Nov 15 20:21:58 2005
@@ -403,14 +403,12 @@
 
  Mason
 *.mas
--autoindent
 -syntax mason
 -smarthome
 -smartbacks
 
  SML
 *.sml
--autoindent
 -syntax sml
 -smarthome
 -smartbacks
@@ -418,7 +416,6 @@
 
  OCaml lanaguage file
 *.ml
--autoindent
 -syntax ocaml
 -smarthome
 -smartbacks
@@ -426,7 +423,6 @@
 
  OCaml language interface
 *.mli
--autoindent
 -syntax ocaml
 -smarthome
 -smartbacks
@@ -435,7 +431,6 @@
  Perl
 *
 +#!\+\[ 	]\+\[a-z/]/perl\>
--autoindent
 -syntax perl
 -smarthome
 -smartbacks
@@ -443,7 +438,6 @@
 -single_quoted
 
 *.pl
--autoindent
 -syntax perl
 -smarthome
 -smartbacks
@@ -451,7 +445,6 @@
 -single_quoted
 
 *.pm
--autoindent
 -syntax perl
 -smarthome
 -smartbacks
@@ -460,7 +453,6 @@
 
  SQL file
 *.sql
--autoindent
 -syntax sql
 -smarthome
 -smartbacks
@@ -473,7 +465,6 @@
 
  AWK language file
 *.awk
--autoindent
 -syntax awk
 -smarthome
 -smartbacks
@@ -482,7 +473,6 @@
 
  YACC
 *.y
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -494,7 +484,6 @@
 
  LEX
 *.l
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -505,7 +494,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.lex
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -518,7 +506,6 @@
  ADA
 *.adb
 -syntax ada
--autoindent
 -istep 2
 -smarthome
 -smartbacks
@@ -528,7 +515,6 @@
 
 *.ads
 -syntax ada
--autoindent
 -istep 2
 -smarthome
 -smartbacks
@@ -556,7 +542,6 @@
 
  C language file
 *.c
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -567,7 +552,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.cpp
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -578,7 +562,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.cc
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -589,7 +572,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.c++
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -601,7 +583,6 @@
 
  C language header file
 *.h
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -612,7 +593,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.hpp
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -623,7 +603,6 @@
 -text_delimiters #if|#ifdef|#ifndef=#elif=#else=#endif
 
 *.h++
--autoindent
 -syntax c
 -smarthome
 -smartbacks
@@ -635,7 +614,6 @@
 
  Verilog file
 *.v
--autoindent
 -syntax verilog
 -istep 2
 -smarthome
@@ -647,7 +625,6 @@
 
  Verilog header file
 *.vh
--autoindent
 -syntax verilog
 -istep 2
 -smarthome
@@ -659,7 +636,6 @@
 
  VHDL file
 *.vhd
--autoindent
 -syntax vhdl
 -istep 2
 -smarthome
@@ -671,29 +647,24 @@
  XML
 *.xml
 -wordwrap
--autoindent
 -syntax xml
 
  XML
 *.xsd
 -wordwrap
--autoindent
 -syntax xml
 
  HTML
 *.htm
 -wordwrap
--autoindent
 -syntax html
 
 *.html
 -wordwrap
--autoindent
 -syntax html
 
  JAVA
 *.java
--autoindent
 -syntax java
 -smarthome
 -smartbacks
@@ -701,7 +672,6 @@
 
  Hypertext preprocessor file
 *.php
--autoindent
 -syntax php
 -smarthome
 -smartbacks
@@ -710,14 +680,12 @@
  Python
 *
 +#!\+\[ 	]\+\[a-z/]/python\>
--autoindent
 -syntax python
 -smarthome
 -smartbacks
 -purify
 
 *.py
--autoindent
 -syntax python
 -smarthome
 -smartbacks
@@ -726,58 +694,48 @@
  CSH
 *
 +#!\+\[ 	]\+\[a-z/]/csh\>
--autoindent
 -syntax csh
 -pound_comment
 
 *.csh
--autoindent
 -syntax csh
 -pound_comment
 
 *.login
--autoindent
 -syntax csh
 -pound_comment
 
 *.logout
--autoindent
 -syntax csh
 -pound_comment
 
  TCSH
 *
 +#!\+\[ 	]\+\[a-z/]/tcsh\>
--autoindent
 -syntax csh
 -pound_comment
 
 *.tcsh
--autoindent
 -syntax csh
 -pound_comment
 
 *.tcshrc
--autoindent
 -syntax csh
 -pound_comment
 
  Shell
 *
 +#!\+\[ 	]\+\[a-z/]/sh\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.sh
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *profile
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
@@ -785,73 +743,61 @@
  BASH Shell
 *
 +#!\+\[ 	]\+\[a-z/]/bash\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash_login
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.bash_logout
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
  LISP
 *.lisp
--autoindent
 -syntax lisp
 -semi_comment
 
 *.lsp
--autoindent
 -syntax lisp
 -semi_comment
 
 *.el
--autoindent
 -syntax lisp
 -semi_comment
 
  KSH
 *
 +#!\+\[ 	]\+\[a-z/]/ksh\>
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
 *.ksh
--autoindent
 -syntax sh
 -pound_comment
 -text_delimiters do=done:if=elif=else=fi:case=esac
 
  Makefile
 *akefile
--autoindent
 -syntax conf
 -pound_comment
 
 *AKEFILE
--autoindent
 -syntax conf
 -pound_comment
 
  Pascal
 *.p
--autoindent
 -syntax pascal
 -smarthome
 -smartbacks
@@ -859,7 +805,6 @@
 -text_delimiters begin|BEGIN|record|RECORD|case|CASE=end|END:repeat|REPEAT=until|UNTIL
 
 *.pas
--autoindent
 -syntax pascal
 -smarthome
 -smartbacks
@@ -868,40 +813,32 @@
 
  Fortran
 *.f
--autoindent
 -syntax fortran
 
 *.for
--autoindent
 -syntax fortran
 
 *.FOR
--autoindent
 -syntax fortran
 
  TCL
 *.tcl
--autoindent
 -syntax tcl
 
  Joe Syntax File
 *.jsf
--autoindent
 -syntax conf
 
  Autoconfig file
 *.ac
--autoindent
 -syntax conf
 
  M4 file
 *.m4
--autoindent
 -syntax m4
 
  Automake file
 *.am
--autoindent
 -syntax conf
 
  Mail file
