$RuOBSD: joe.patch,v 1.1 2005/11/15 14:34:22 form Exp $

--- utf8.c.orig	Sun Mar 18 10:24:23 2007
+++ utf8.c	Sun Mar 18 10:24:37 2007
@@ -345,12 +345,8 @@
 	unsigned char *s, *t, *u;
 
 	s=(unsigned char *)getenv("LC_ALL");
-	if (!s) {
-		s=(unsigned char *)getenv("LC_CTYPE");
-		if (!s) {
-			s=(unsigned char *)getenv("LANG");
-		}
-	}
+	if (!s)
+		s=(unsigned char *)getenv("LANG");
 
 	if (s)
 		s=zdup(s);
