$RuOBSD: mount.patch,v 1.1 2003/11/17 06:38:58 form Exp $

--- mount.c.orig	Fri Nov 14 10:37:56 2003
+++ mount.c	Fri Nov 14 10:38:38 2003
@@ -92,7 +92,7 @@
 	const char *o_name;
 	const char *o_optname;
 } optnames[] = {
-	{ MNT_ASYNC,		0,	"asynchronous",		"async" },
+	{ MNT_ASYNC,		0,	"async",		"async" },
 	{ MNT_DEFEXPORTED,	1,	"exported to the world", "" },
 	{ MNT_EXKERB,		1,	"kerberos uid mapping",	"" },
 	{ MNT_EXPORTED,		0,	"NFS exported",		"" },
@@ -100,7 +100,7 @@
 	{ MNT_EXRDONLY,		1,	"exported read-only",	"" },
 	{ MNT_LOCAL,		0,	"local",		"" },
 	{ MNT_NOATIME,		0,	"noatime",		"noatime" },
-	{ MNT_NOATIME,		0,	"noaccesstime",		"" },
+	{ MNT_NOATIME,		0,	"noatime",		"" },
 	{ MNT_NODEV,		0,	"nodev",		"nodev" },
 	{ MNT_NOEXEC,		0,	"noexec",		"noexec" },
 	{ MNT_NOSUID,		0,	"nosuid",		"nosuid" },
@@ -107,7 +107,7 @@
 	{ MNT_QUOTA,		0,	"with quotas",		"" },
 	{ MNT_RDONLY,		0,	"read-only",		"ro" },
 	{ MNT_ROOTFS,		1,	"root file system",	"" },
-	{ MNT_SYNCHRONOUS,	0,	"synchronous",		"sync" },
+	{ MNT_SYNCHRONOUS,	0,	"sync",			"sync" },
 	{ MNT_SOFTDEP,		0,	"softdep", 		"softdep" },
 	{ MNT_UNION,		0,	"union",		"" },
 	{ NULL,			0,	"",			"" }
