$RuOBSD$

--- VAX/vax780_sbi.c.orig	Thu Mar 16 01:16:49 2006
+++ VAX/vax780_sbi.c	Thu Mar 16 01:17:57 2006
@@ -30,6 +30,10 @@
 
 #include "vax_defs.h"
 
+#ifndef PATH_VMB_EXE
+#define PATH_VMB_EXE	"vmb.exe"
+#endif
+
 /* 11/780 specific IPRs */
 
 /* Writeable control store */
@@ -635,10 +639,10 @@ t_stat cpu_boot (int32 unitno, DEVICE *d
 {
 t_stat r;
 
-printf ("Loading boot code from vmb.exe\n");
+printf ("Loading boot code from %s\n", PATH_VMB_EXE);
 if (sim_log) fprintf (sim_log, 
-    "Loading boot code from vmb.exe\n");
-r = load_cmd (0, "-O vmb.exe 200");
+    "Loading boot code from %s\n", PATH_VMB_EXE);
+r = load_cmd (0, "-O " PATH_VMB_EXE " 200");
 if (r != SCPE_OK) return r;
 SP = PC = 512;
 return SCPE_OK;
