$RuOBSD$

--- display.c.orig	Tue Mar 29 23:27:54 2005
+++ display.c	Tue Mar 29 23:28:11 2005
@@ -46,6 +46,7 @@
  */
 
 #include <sys/types.h>
+#include <math.h>
 #include <stdio.h>
 #include <ctype.h>
 #include <stdlib.h>
@@ -53,6 +54,7 @@
 #include <signal.h>
 #include <term.h>
 #include <time.h>
+#include <tzfile.h>
 #include <unistd.h>
 #include <stdarg.h>
 
@@ -217,6 +219,34 @@
 	/* we should optimize this and only display changes */
 	for (i = 0; i < 3; i++)
 		printf("%s%5.2f", i == 0 ? "" : ", ", avenrun[i]);
+}
+
+void
+i_uptime(struct timeval* bt, time_t *tod)
+{
+	time_t uptime;
+	int days, hrs, mins, secs;
+
+	if (bt->tv_sec != -1) {
+		uptime = *tod - bt->tv_sec;
+		uptime += 30;
+		days = uptime / SECSPERDAY;
+		uptime %= SECSPERDAY;
+		hrs = uptime / SECSPERHOUR;
+		uptime %= SECSPERHOUR;
+		mins = uptime / SECSPERMIN;
+		secs = uptime % SECSPERMIN;
+
+		/*
+		 *  Display the uptime.
+		 */
+		if (smart_terminal)
+			Move_to((screen_width - 24) - (days == 0 ? 0 :
+			    (int)log10f((float)days)), 0);
+		else
+			fputs(" ", stdout);
+		printf(" up %d+%02d:%02d:%02d", days, hrs, mins, secs);
+	}
 }
 
 /*
--- machine.c.orig	Tue Dec  7 16:40:23 2004
+++ machine.c	Tue Mar 29 21:13:51 2005
@@ -217,6 +217,7 @@
 	static int sysload_mib[] = {CTL_VM, VM_LOADAVG};
 	static int vmtotal_mib[] = {CTL_VM, VM_METER};
 	static int cp_time_mib[] = {CTL_KERN, KERN_CPTIME};
+	static int boottime_mib[] = {CTL_KERN, KERN_BOOTTIME};
 	struct loadavg sysload;
 	struct vmtotal vmtotal;
 	double *infoloadp;
@@ -260,6 +261,10 @@
 	si->cpustates = cpu_states;
 	si->memory = memory_stats;
 	si->last_pid = -1;
+
+	size = sizeof(si->boottime);
+	if (sysctl(boottime_mib, 2, &si->boottime, &size, NULL, 0) < 0)
+		si->boottime.tv_sec = -1;
 }
 
 static struct handle handle;
--- machine.h.orig	Tue Oct 12 10:52:43 2004
+++ machine.h	Tue Mar 29 21:15:03 2005
@@ -57,6 +57,7 @@
 	int            *procstates;
 	int            *cpustates;
 	int            *memory;
+	struct timeval	boottime;
 };
 
 /*
--- top.c.orig	Tue Oct 12 10:52:43 2004
+++ top.c	Tue Mar 29 21:05:24 2005
@@ -409,9 +409,10 @@
 		/* display the load averages */
 		(*d_loadave)(system_info.last_pid, system_info.load_avg);
 
-		/* display the current time */
+		/* display the current time and uptime */
 		/* this method of getting the time SHOULD be fairly portable */
 		time(&curr_time);
+		i_uptime(&system_info.boottime, &curr_time);
 		i_timeofday(&curr_time);
 
 		/* display process state breakdown */
