$RuOBSD: top.patch,v 1.4 2007/05/30 08:02:29 form Exp $

--- display.c.orig	Fri Oct 12 05:55:41 2007
+++ display.c	Fri Oct 12 05:55:57 2007
@@ -55,6 +55,7 @@
 #include <signal.h>
 #include <stdlib.h>
 #include <string.h>
+#include <tzfile.h>
 #include <unistd.h>
 
 #include "screen.h"		/* interface to screen package */
@@ -268,6 +269,39 @@
 	putn();
 }
 
+/*
+ * Display the uptime
+ */
+void
+i_uptime(struct timeval *bt, time_t *tod)
+{
+	time_t uptime;
+	int days, hrs, mins, secs;
+
+	if (bt->tv_sec != -1) {
+		uptime = *tod - bt->tv_sec;
+		uptime += 30;
+		days = uptime / SECSPERDAY;
+		uptime %= SECSPERDAY;
+		hrs = uptime / SECSPERHOUR;
+		uptime %= SECSPERHOUR;
+		mins = uptime / SECSPERMIN;
+		secs = uptime % SECSPERMIN;
+
+		/*
+		 *  Display the uptime.
+		 */
+		if (smart_terminal)
+			move(0, (screen_width - 24) - (days > 9) -
+			    (days > 99) - (days > 999) - (days > 9999));
+		else {
+			if (fputs(" ", stdout) == EOF)
+				exit(1);
+		}
+		printwp(" up %d+%02d:%02d:%02d", days, hrs, mins, secs);
+	}
+}
+ 
 /*
  *  *_procstates(total, brkdn, names) - print the process summary line
  *
--- display.h.orig	Wed May 30 14:17:12 2007
+++ display.h	Wed May 30 14:17:44 2007
@@ -37,6 +37,7 @@
 int display_resize(void);
 void i_loadave(int, double *);
 void u_loadave(int, double *);
+void i_uptime(struct timeval *, time_t *);
 void i_timeofday(time_t *);
 void i_procstates(int, int *);
 void u_procstates(int, int *);
--- machine.c.orig	Sun Jun 19 00:35:03 2005
+++ machine.c	Sun Jun 19 00:38:14 2005
@@ -232,6 +232,7 @@
 {
 	static int sysload_mib[] = {CTL_VM, VM_LOADAVG};
 	static int vmtotal_mib[] = {CTL_VM, VM_METER};
+	static int boottime_mib[] = {CTL_KERN, KERN_BOOTTIME};
 	struct loadavg sysload;
 	struct vmtotal vmtotal;
 	double *infoloadp;
@@ -295,6 +296,10 @@
 	si->cpustates = cpu_states;
 	si->memory = memory_stats;
 	si->last_pid = -1;
+
+	size = sizeof(si->boottime);
+	if (sysctl(boottime_mib, 2, &si->boottime, &size, NULL, 0) < 0)
+		si->boottime.tv_sec = -1;
 }
 
 static struct handle handle;
--- machine.h.orig	Sun Jun 19 00:35:03 2005
+++ machine.h	Sun Jun 19 00:35:03 2005
@@ -57,6 +57,7 @@
 	int            *procstates;
 	int64_t        *cpustates;
 	int            *memory;
+	struct timeval	boottime;
 };
 
 /*
--- top.c.orig	Sun Jun 19 00:35:03 2005
+++ top.c	Sun Jun 19 00:35:03 2005
@@ -412,9 +412,10 @@
 		/* display the load averages */
 		(*d_loadave)(system_info.last_pid, system_info.load_avg);
 
-		/* display the current time */
+		/* display the current time and uptime */
 		/* this method of getting the time SHOULD be fairly portable */
 		time(&curr_time);
+		i_uptime(&system_info.boottime, &curr_time);
 		i_timeofday(&curr_time);
 
 		/* display process state breakdown */
