Index: conf.c
===================================================================
RCS file: /cvs/src/sys/arch/i386/i386/conf.c,v
retrieving revision 1.80
diff -u -r1.80 conf.c
--- conf.c	2001/09/28 02:53:13	1.80
+++ conf.c	2001/10/03 06:30:24
@@ -218,6 +218,8 @@
 cdev_decl(ucom);
 #include "cz.h"
 cdev_decl(cztty);
+#include "radio.h"
+cdev_decl(radio);
 
 /* XXX -- this needs to be supported by config(8)! */
 #if (NCOM > 0) && (NPCCOM > 0)
@@ -345,6 +347,7 @@
 	cdev_pf_init(NPF,pf),		/* 73: packet filter */
 	cdev_altq_init(NALTQ,altq),	/* 74: ALTQ control interface */
 	cdev_iop_init(NIOP,iop),	/* 75: I2O IOP control interface */
+	cdev_radio_init(NRADIO, radio), /* 76: generic radio I/O */
 };
 int	nchrdev = sizeof(cdevsw) / sizeof(cdevsw[0]);
 
