$RuOBSD$

--- src/getname.c.orig	Thu Feb 13 00:07:53 2003
+++ src/getname.c	Fri Mar  7 09:49:59 2003
@@ -16,25 +16,9 @@
 
 char *spamify(char *input)
 {
-	int insertlen = strlen(set_antispam_at);
-  /* we should replace the @-letter in the email
-     address */
-	int newlen = strlen(input) + insertlen;
 	char *atptr = strchr(input, '@');
-	if (atptr) {
-    char *newbuf = malloc(newlen);
-		int index = atptr - input;
-    /* copy the part before the @ */
-    memcpy(newbuf, input, index);
-
-		memcpy(newbuf + index, set_antispam_at, insertlen);
-    /* append the part after the @ */
-		strcpy(newbuf + index + insertlen, input + index + 1);
-    /* correct the pointer and free the old */
-    free(input);
-    return newbuf;
-  }
-  /* weird email, bail out */
+	if (atptr)
+		*atptr = 0;
   return input;
 }
 
