$RuOBSD: locale-hack.patch,v 1.1 2004/01/12 03:31:52 form Exp $

--- setlocale.c.orig	Sun Jan 11 22:44:50 2004
+++ setlocale.c	Sun Jan 11 23:05:17 2004
@@ -63,10 +63,10 @@
 static char current_categories[_LC_LAST][32] = {
     "C",
     "C",
+    "ru_RU.KOI8-R",
     "C",
     "C",
     "C",
-    "C",
     "C"
 };
 
@@ -209,7 +209,8 @@
 	    current_categories[category]) == 0)
 		return (current_categories[category]);
 
-	if (!strcmp(new_categories[category], "C") ||
+	if (!strcmp(new_categories[category], "ru_RU.KOI8-R") ||
+	    !strcmp(new_categories[category], "C") ||
 	    !strcmp(new_categories[category], "POSIX")) {
 
 		/*
@@ -261,9 +262,9 @@
 	if (!env || !*env)
 		env = getenv("LANG");
 
-	/* 4. if none is set, fall to "C" */
+	/* 4. if none is set, fall to "ru_RU.KOI8-R" */
 	if (!env || !*env || strchr(env, '/'))
-		env = "C";
+		env = "ru_RU.KOI8-R";
 
 	return env;
 }
