$RuOBSD$

--- pptpctrl.c.orig	Fri Mar 19 22:19:25 2004
+++ pptpctrl.c	Fri Mar 19 22:17:08 2004
@@ -191,6 +191,8 @@
 	NOTE_VALUE(PAC, call_id_pair, htons(-1));
 	NOTE_VALUE(PNS, call_id_pair, htons(-1));
 
+	setenv("PPTP_CLIENT", inet_ntoa(addr.sin_addr), 1);
+
 	syslog(LOG_INFO, "CTRL: Client %s control connection started", inet_ntoa(addr.sin_addr));
 	pptp_handle_ctrl_connection(pppaddrs, inetaddrs);
 #ifdef BCRELAY
