--- files.isa.orig	Wed Jan 31 01:57:39 2001
+++ files.isa	Sun Oct  7 14:20:16 2001
@@ -352,3 +352,23 @@
 #file dev/isa/pcmcia_pcic.c  pcic | pcicmaster
 
 #file dev/isa/pcmcia_isa.c  pcmcia
+
+# Sound Forte RadioLink SF16-FMR2 FM Radio Card
+device sf2r: radio, tea5757
+attach sf2r at isa
+file	dev/isa/sf16fmr2.c	sf2r
+
+# Aztech/PackardBell FM Radio Card
+device az: radio, lm700x
+attach az at isa
+file	dev/isa/aztech.c	az
+
+# AIMS Lab Radiotrack & compatible
+device rt: radio, lm700x
+attach rt at isa
+file	dev/isa/radiotrack.c	rt
+
+# AIMS Lab Radiotrack II FM Radio Card
+device rtii: radio, tea5757
+attach rtii at isa
+file	dev/isa/radiotrack2.c	rtii
